windows/src/desktop/kmshell/main/Keyman.System.UpdateStateMachine.pas

```diff
@@ -0,0 +1,1348 @@
+      end;
+    end;
+  finally
+    Registry.Free;
+  end;
+end;
+
+function TUpdateStateMachine.GetApplyNow: Boolean;
+var
+  Registry: TRegistryErrorControlled;
+begin
+  // check the registry value
+  Registry := TRegistryErrorControlled.Create;
+  try
+    Registry.RootKey := HKEY_CURRENT_USER;
+    Result := Registry.OpenKeyReadOnly(SRegKey_KeymanEngine_CU) and
+      Registry.ValueExists(SRegValue_ApplyNow) and
+      Registry.ReadBool(SRegValue_ApplyNow);
+  finally
+    Registry.Free;
+  end;
+end;
+
+
+function TUpdateStateMachine.CheckUpdateSchedule: Boolean;
+var
+  RegistryErrorControlled :TRegistryErrorControlled;
+begin
+  try
+    Result := False;
+    RegistryErrorControlled := TRegistryErrorControlled.Create;
+
+    try
+      if RegistryErrorControlled.OpenKeyReadOnly(SRegKey_KeymanDesktop_CU) then
+      begin
+        if RegistryErrorControlled.ValueExists(SRegValue_CheckForUpdates) and not RegistryErrorControlled.ReadBool(SRegValue_CheckForUpdates) and not FForce then
+        begin
+          Result := False;
+          Exit;
+        end;
+        if RegistryErrorControlled.ValueExists(SRegValue_LastUpdateCheckTime) and (Now - RegistryErrorControlled.ReadDateTime(SRegValue_LastUpdateCheckTime) < 1) and not FForce then
+        begin
+          Result := False;
+          Exit;
+        end;
+        // Else Time to check for updates
+        Result := True;
+      end;
+    finally
+      RegistryErrorControlled.Free;
+    end;
+  except
+    { we will not run the check if an error occurs reading the settings }
+    on E:Exception do
+    begin
+      Result := False;
+      FErrorMessage := E.Message;
+      Exit;
+    end;
+  end;
+end;
+
+function TUpdateStateMachine.GetState: TStateClass;
+begin
+  Result := TStateClass(CurrentState.ClassType);
+end;
+
+procedure TUpdateStateMachine.SetState(const Value: TStateClass);
+begin
+  if Assigned(CurrentState) then
+  begin
+    CurrentState.Exit;
+  end;
+
+  SetStateOnly(Value);
+
+  if Assigned(CurrentState) then
+  begin
+    CurrentState.Enter;
+  end
+  else
+  begin
+   // TODO: #10210 Error log for Unable to set state for Value
+  end;
+
+end;
+
+procedure TUpdateStateMachine.SetStateOnly(const Value: TStateClass);
+begin
+  if Value = IdleState then
+  begin
+    CurrentState := FIdle;
+  end
+  else if Value = UpdateAvailableState then
+  begin
+    CurrentState := FUpdateAvailable;
+  end
+  else if Value = DownloadingState then
+  begin
+    CurrentState := FDownloading;
+  end
+  else if Value = WaitingRestartState then
+  begin
+    CurrentState := FWaitingRestart;
+  end
+  else if Value = InstallingState then
+  begin
+    CurrentState := FInstalling;
+  end
+  else if Value = RetryState then
+  begin
+    CurrentState := FRetry;
+  end
+  else if Value = PostInstallState then
+  begin
+    CurrentState := FPostInstall;
+  end;
+end;
+
+function TUpdateStateMachine.ConvertEnumState(const TEnumState: TUpdateState) : TStateClass;
```

Comment:
```suggestion
function TUpdateStateMachine.ConvertEnumState(const updateState: TUpdateState) : TStateClass;
```

`T` prefix is for types, not variables or parameters.

This function would not be required if we used the array map for state instances as noted above